hbase-server/src/test/java/org/apache/hadoop/hbase/security/provider/TestCustomSaslAuthenticationProvider.java

```diff
@@ -0,0 +1,539 @@
+
+  private static final String USER1_PASSWORD = "foobarbaz";
+  private static final String USER2_PASSWORD = "bazbarfoo";
+
+  private static Map<String,String> createUserDatabase() {
+    Map<String,String> db = new ConcurrentHashMap<>();
+    db.put("user1", USER1_PASSWORD);
+    db.put("user2", USER2_PASSWORD);
+    return db;
+  }
+
+  public static String getPassword(String user) {
+    String password = USER_DATABASE.get(user);
+    if (password == null) {
+      throw new IllegalStateException("Cannot request password for a user that doesn't exist");
+    }
+    return password;
+  }
+
+  /**
+   * A custom tokenidentifier for our custom auth'n method. Unique from the TokenIdentifier
+   * used for delegation tokens.
+   */
+  public static class PasswordAuthTokenIdentifier extends TokenIdentifier {
+    public static final Text PASSWORD_AUTH_TOKEN = new Text("HBASE_PASSWORD_TEST_TOKEN");
+    private String username;
+
+    public PasswordAuthTokenIdentifier() {}
+
+    public PasswordAuthTokenIdentifier(String username) {
+      this.username = username;
+    }
+
+    @Override
+    public void readFields(DataInput in) throws IOException {
+      this.username = WritableUtils.readString(in);
+    }
+
+    @Override
+    public void write(DataOutput out) throws IOException {
+      WritableUtils.writeString(out, username);
+    }
+
+    @Override
+    public Text getKind() {
+      return PASSWORD_AUTH_TOKEN;
+    }
+
+    @Override
+    public UserGroupInformation getUser() {
+      if (username == null || "".equals(username)) {
+        return null;
+      }
+      return UserGroupInformation.createRemoteUser(username);
+    }
+  }
+
+  public static Token<? extends TokenIdentifier> createPasswordToken(
+      String username, String password, String clusterId) {
+    PasswordAuthTokenIdentifier id = new PasswordAuthTokenIdentifier(username);
+    Token<? extends TokenIdentifier> token = new Token<>(id.getBytes(), Bytes.toBytes(password),
+        id.getKind(), new Text(clusterId));
+    return token;
+  }
+
+  /**
+   * Client provider that finds custom Token in the user's UGI and authenticates with the server
+   * via DIGEST-MD5 using that password.
+   */
+  public static class InMemoryClientProvider extends AbstractSaslClientAuthenticationProvider {
+    public static final String MECHANISM = "DIGEST-MD5";
+    public static final SaslAuthMethod SASL_AUTH_METHOD = new SaslAuthMethod(
+        "IN_MEMORY", (byte)42, MECHANISM, AuthenticationMethod.TOKEN);
+
+    @Override
+    public SaslClient createClient(Configuration conf, InetAddress serverAddr,
+        SecurityInfo securityInfo, Token<? extends TokenIdentifier> token, boolean fallbackAllowed,
+        Map<String, String> saslProps) throws IOException {
+      return Sasl.createSaslClient(new String[] { MECHANISM }, null, null,
+          SaslUtil.SASL_DEFAULT_REALM, saslProps, new InMemoryClientProviderCallbackHandler(token));
```

Comment:
Possibly, the failure mentioned in comment (https://github.com/apache/hbase/pull/884#discussion_r358563326) is because of this SASL_DEFAULT_REALM?

Reply:
Yeah, that was it. I didn't realize that we passed in a default realm here.